big_vision/trainers/proj/gsam/train.py

```diff
@@ -0,0 +1,401 @@
+      config.dataset, config.train_split,
+      data_dir=fillin(config.get("dataset_dir")))
+  steps_per_epoch = ntrain_img / batch_size
+
+  if config.get("num_epochs"):
+    total_steps = int(config.num_epochs * steps_per_epoch)
+    assert not config.get("total_steps"), "Set either num_epochs or total_steps"
+  else:
+    total_steps = config.total_steps
+
+  info("Running for %d steps, that means %f epochs and %f steps per epoch",
+       total_steps, total_steps * batch_size / ntrain_img, steps_per_epoch)
+
+  write_note(f"Initializing {config.model_name} model...")
+  model_mod = importlib.import_module(f"big_vision.models.{config.model_name}")
+  model = model_mod.Model(
+      num_classes=config.num_classes, **config.get("model", {}))
+
+  # We want all parameters to be created in host RAM, not on any device, they'll
+  # be sent there later as needed, otherwise we already encountered two
+  # situations where we allocate them twice.
+  @partial(jax.jit, backend="cpu")
+  def init(rng):
+    shape = tuple(train_ds.element_spec["image"].shape[1:])
+    bs = config.batch_size // jax.device_count()
+    dummy_input = jnp.zeros((bs,) + shape, jnp.float32)
+    params = flax.core.unfreeze(model.init(rng, dummy_input))["params"]
+
+    # Set bias in the head to a low value, such that loss is small initially.
+    if "init_head_bias" in config:
+      params["head"]["bias"] = jnp.full_like(params["head"]["bias"],
+                                             config["init_head_bias"])
+
+    return params
+
+  rng, rng_init = jax.random.split(rng)
+  params_cpu = init(rng_init)
+
+  if jax.process_index() == 0:
+    num_params = sum(p.size for p in jax.tree_leaves(params_cpu))
+    parameter_overview.log_parameter_overview(params_cpu, msg="init params")
+    mw.measure("num_params", num_params)
+
+  write_note(f"Initializing {config.optax_name} optimizer...")
+  tx, sched_fns = bv_optax.make(config, params_cpu, sched_kw=dict(
+      global_batch_size=batch_size,
+      total_steps=total_steps,
+      steps_per_epoch=steps_per_epoch))
+
+  # We jit this, such that the arrays are created on the CPU, not device[0].
+  opt_cpu = jax.jit(tx.init, backend="cpu")(params_cpu)
+  sched_fns_cpu = [jax.jit(sched_fn, backend="cpu") for sched_fn in sched_fns]
+
+  # ----------------------------------------------------------------------------
+  # Get hyper-params for GSAM.
+  RHO_MAX, RHO_MIN = config.get("rho_max", 0.3), config.get("rho_min", 0.05)
+
+  warnings.warn("Please make sure 'config.lr_min' equals the min value by your lr_schedule. \
+    Perturbation amplitude schedule is: \
+    \n sam_rho = rho_min + (rho_max - rho_min) * (lr - lr_min) / (lr_max - lr_min) ")
+
+  LR_MAX = config.get("lr", 3e-3)
+  LR_MIN = config.get("lr_min", LR_MAX * 0.01)
+
+  ALPHA = config.get("alpha", 0.05)
+  ADAPTIVE_PERTURBATION = config.get("adaptive_perturbation", False)
+  MINIMIZE_FP = config.get("minimize_fp", True)
```

Comment:
Each of these is actually only used exactly once below, so in our code style, we would not assign them to any variable, but just inline them where they are used, see comment below.